libraries/chain/controller.cpp

```diff
@@ -3999,18 +3985,138 @@ struct controller_impl {
       bsp->verify_qc(qc_proof);
    }
 
+   void verify_legacy_block_exts( const signed_block_ptr& b, const block_header_state_legacy& prev ) {
+      assert(b->is_legacy_block());
+
+      uint32_t block_num = b->block_num();
+      auto block_exts = b->validate_and_extract_extensions();
+      auto qc_ext_id = quorum_certificate_extension::extension_id();
+      bool qc_extension_present = block_exts.count(qc_ext_id) != 0;
+
+      EOS_ASSERT( !qc_extension_present, invalid_qc_claim,
+                  "Legacy block #${b} includes a QC block extension",
+                  ("b", block_num) );
+
+      EOS_ASSERT( !b->is_proper_svnn_block(), invalid_qc_claim,
+                  "Legacy block #${b} has invalid schedule_version",
+                  ("b", block_num) );
+
+      // Verify we don't go back from Savanna (Transition or Proper) block to Legacy block
+      auto it = prev.header_exts.find(finality_extension::extension_id());
+      EOS_ASSERT( it == prev.header_exts.end(), invalid_qc_claim,
+                  "Legacy block #${b} has previous block which contains finality block header extension",
+                  ("b", block_num) );
+      EOS_ASSERT( !prev.header.is_proper_svnn_block(), invalid_qc_claim,
+                  "Legacy block #${b} may not have previous block that is a Proper Savanna block",
+                  ("b", block_num) );
+   }
+
+   void verify_transition_block_exts( const signed_block_ptr& b, const block_header_state_legacy& prev ) {
+      assert(!b->is_legacy_block());
+
+      uint32_t block_num = b->block_num();
+      auto block_exts = b->validate_and_extract_extensions();
+      auto qc_ext_id = quorum_certificate_extension::extension_id();
+      bool qc_extension_present = block_exts.count(qc_ext_id) != 0;
+
+      EOS_ASSERT( !qc_extension_present, invalid_qc_claim,
+                  "Transition block #${b} includes a QC block extension",
+                  ("b", block_num) );
+
+      EOS_ASSERT( !b->is_proper_svnn_block(), invalid_qc_claim,
```

Comment:
This check is redundant given it is part of the definition of the Transition block and therefore must be true before the caller of this function decides it is even appropriate to pass block `b` to `verify_transition_block_exts`. So it should be removed.

But to detect programmer errors in DEBUG mode, it is important to make the modification I noted to the `assert` above:

```cpp
assert(!b->is_legacy_block() && !b->is_proper_svnn_block());
```